Tutorial: setting up PostHog for multiple environments (#4930)
PostHog/posthog.com: contents/tutorials/multiple-environments.md

---
title: How to setup PostHog for multiple environments (development, staging, production)
sidebar: Docs
showTitle: true
author: ['ian-vanagas']
date: 2023-01-02
featuredImage: ../images/tutorials/banners/multiple-environments.png
topics: ["configuration"]
---

Many software development teams use environments to split up their code, such as development, staging, and production. This ensures changes in development don’t affect production, helps teams test before release, and increases the quality of code that reaches end users.

Using multiple environments requires splitting the data from each of them. If not, development and staging data combine with production data and can cause inaccuracy and issues. This tutorial goes over how to prevent those issues by setting up PostHog for use in multiple environments.

## Using multiple projects

The best practice for using PostHog with multiple environments is using multiple projects. This enables developers to test event capture, session recordings, and feature flags without polluting the data in production.

> Creating a new project in PostHog is simple. In your instance, click your project name and then the “New project” button. Be sure to name your project to make it clear it is not production, like “development” or “staging.”

Reviewer: ```suggestion
Creating a new project in PostHog is simple. In your instance, click your project name and then the “New project” button. Be sure to name your project to make it clear it is not production, like “development” or “staging”.
```

Author: I like to keep my punctuation inside the quotes 🤔

Reviewer: I don't, unless it's speech, but I don't have a strong feeling at all haha

![Projects](../images/tutorials/multiple-environments/project.png)

> This takes you through the project setup flow again and gives you a new project API key. You can keep your old key for production, and add the new one in the relevant environment. You'll likely need environment variables for this (like a `.env` file) if you haven't set those up. Initializing PostHog with a different key in each environment connects them to a different project.

Reviewer: Should we cover how to set up env variables here? The phrasing suggests I should expect it to be covered, but it's a bit out of my technical understanding

Author: Added it

> The downside of using multiple projects is you cannot directly copy actions, dashboards, insights, experiments, feature flags, and other data created in PostHog between them. This means manually recreating them in each project if needed.

Reviewer: ```suggestion
The downside of using multiple projects is that you cannot directly copy actions, dashboards, insights, experiments, feature flags, and other data created in PostHog between them. This means you must manually recreate them in each project if needed.
```

*Have strong opinions about how we handle environments? We’d love to get your support and feedback in the [relevant issue](https://github.com/PostHog/posthog/issues/13418).*

### Feature flags with multiple projects

> Testing feature flags in development and staging environments ensure they are working properly, but you must recreate the flags in each project to use them in separate environments. 

Reviewer: ```suggestion
Testing feature flags in development and staging environments ensures they are working properly, but you must recreate the flags in each project to use them in separate environments. 
```

> Luckily, creating a new flag is relatively simple. Make sure to use the same flag keys and variants between the projects, and the flag behaves the same. You also don’t have to worry about rollout percentages for test or staging environments (you can roll out to everyone on your team).

Reviewer: ```suggestion
Luckily, creating a new flag is simple. Make sure to use the same flag keys and variants between the projects, and the flag will behave the same across each environment. You also don’t have to worry about roll out percentages for test or staging environments (you can roll out to everyone on your team).
```

*We are also gathering feedback on how to use feature flags with multiple projects. Leave your support and feedback in [the issue here](https://github.com/PostHog/posthog/issues/13160).*

## Conditional initialization

> 

Reviewer: ```suggestion
### Token-based opt-out
```

Another way to split up projects is by changing how you initialize PostHog based on the environment. For example, if you didn’t want to track your staging environment at all, initialize PostHog without a token and opt out of capturing immediately. This is what we do in our development environment.

```js
posthog.init(
    'fake token',
    {
      autocapture: false,
      loaded: function (ph) {
          ph.opt_out_capturing()
      }, 
    }
  )
```

> The loaded method lets you run whatever checks you want, making it useful for changing PostHog's behavior between environments. You can see the full list of config options in [our JavaScript docs](/docs/integrate/client/js#config).

Reviewer: ```suggestion
This method enable you to run whatever checks you want, making it useful for changing PostHog's behavior between environments. You can see the full list of config options in [our JavaScript docs](/docs/integrate/client/js#config).
```

> 

Reviewer: ```suggestion
### URL-based opt-out
```

You can probably think of a better title, but I think worth labelling these clearly

> Another popular option is checking if the URL includes `localhost` or `127.0.0.1` (local IP) and opting out of capturing if so. If you were using the HTML snippet, this looks like:

Reviewer: ```suggestion
Another popular option is checking if the URL includes `localhost` or `127.0.0.1` (local IP) and opting out of capturing if so. If you were using the HTML snippet, this would look like:
```

	  posthog.init('<ph_project_api_key>',{api_host:'<ph_instance_address>'})
	  if (location.hostname === "localhost" || location.hostname === "127.0.0.1") {
	    posthog.opt_out_capturing()
	  }
</script>
```

Setting this up correctly prevents capturing non-production event data. It enables you to continue using the same insights, feature flags, and other PostHog tools in development and staging because you only use a single project for all your data.

> > **Note:** If you are using multiple projects, be careful not to opt out of capture unnecessarily. You can instead check for the environment and tune autocapture to capture the right data. You can learn about autocapture tuning in [our docs](https://posthog.com/docs/integrate/client/js#tuning-autocapture).

Reviewer: ```suggestion
> **Note:** If you are using multiple projects, be careful not to opt out of capture unnecessarily. You can instead check for the environment and [tune autocapture to capture the right data](/docs/integrate/client/js#tuning-autocapture).
```

## Filtering internal and test users

The previous options capture data to separate projects or don’t capture data at all, but there is another option. This is to capture data normally and filter internal users from your analysis.

> PostHog provides a toggle to filter internal users (as defined by you) from your analysis and visualization. To set up what users or events get filtered, go to Project Settings, scroll down to “Filter out internal and test users,” and add filters removing your internal users and events. This could include filters like: 

Reviewer: ```suggestion
PostHog provides a toggle to filter internal users (as defined by you) from your analysis and visualization. To set up what users or events get filtered, go to Project Settings, scroll down to “Filter out internal and test users,” and add filters to identify your internal users and events, so that they can be removed. This could include filters like: 

```
